### PR TITLE
Moves Game

### LINKED ISSUE #106: Moves Game
**Is your feature request related to a problem? Please describe.**
The user is not able to fight with the enemy pokemon.

**Describe the solution you'd like**
The user should be able to pick a move from the list and fight it against the enemy pokemon, and the same thing should happen with the enemy pokemon fighting back afterwards.

### PR DESCRIPTION
**Pull request by BlackCubes**
## Changes
1. Created a `damageEffectiveness` function in `effectiveness.js` which calculates the damage dealt when a Pokémon uses a damaging move.
2. Created a `ProgressBar` component to display the health of both pokemons when they are dealing damage, and styled it.
3. Created a handler function in `SinglePokemon.js` and `EnemyPokemon.js` to calculate the damage, and passed the handler down to `MovesList.js` to run when the user selects a move.
4. Created an `updateHPStats` utility function that updates the pokemon's HP by decrementing from the damage if attacking, or add the original HP back.

## Purpose
Not really a problem, but it would be cool if two pokemons duke it out.

## Approach
To calculate the damage, research was done to implement a calculation based on [Bulbapedia](https://bulbapedia.bulbagarden.net/wiki/Damage)'s article. Afterwards, a `damageEffectiveness` was created to closely resemble the game.

Next, a handler function called `handleDamageEffectiveness` was implemented on both `SinglePokemon.js` and `EnemyPokemon.js`, and passed down to `MovesList.js` to calculate the damage that should be given after the user selects a move.

To update the pokemon's HP from the damage, a utility function called `updateHPStats` was created that decrements the pokemon's previous HP. To dynamically render the new HP after taking damage, two states called `friendlyPokeStats` and `enemyPokeStats` were created and used. The `updateHPStats` utility function also keeps an HP of 0 no matter what so that there would be no negative values. Lastly, this utility function also resets the HP back to the pokemon's original HP whenever the user finds a new pokemon, or goes back to the single screen.

A `ProgressBar` component was also created to dynamically show the pokemon's HP changing with three different colors: Green, Yellow, and Red.

Overall, this creates a lit game.

## Learning
[Bulbapedia: Damage](https://bulbapedia.bulbagarden.net/wiki/Damage)

Closes #106

### CODE REVIEW

File: pokemon-app/src/components/ProgressBar.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+const ProgressBar = ({ progress }) => {
+  const [style, setStyle] = useState({});
+
+  useEffect(() => {
+    const hpColorStatus =
+      progress > 50
+        ? "#00ae00"
+        : progress > 20 && progress <= 50
+        ? "yellow"
+        : "red";
```
Comment (kmausolf) [inline pokemon-app/src/components/ProgressBar.js:14]: Niiiice, you even set the color of the HP bar!

File: pokemon-app/src/utils/effectiveness.js [hunk 2]
```diff
@@ -133,14 +140,149 @@ const moveEffectiveness = (defenderType, attackerMoveTypes) => {
...
+    critical *
+    stab *
+    type *
+    burn *
+    other;
+
+  // A random factor for the damage: From Generation III onward, it is a random
+  // integer percentage between 0.85 and 1.00 (inclusive). In Generations I and II,
+  // it is realized as a multiplication by a random uniformly distributed integer
+  // between 217 and 255 (inclusive), followed by an integer division by 255.
```
Comment (kmausolf) [inline pokemon-app/src/utils/effectiveness.js:273]: Huh, I had no idea that they changed it!

File: pokemon-app/src/utils/randomGenerator.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  // simply add the min after multiplying Math.round() with (max - min).
+  // Next, to get an even distribution is to do Math.floor().
+  // YET, what if you wanted max to be INCLUSIVE?
+  // Since we are limited with the boundary condition of [min, max) because of the
+  // interval of Math.round() [0, 1), the simple fix is to simply add 1 to the max
+  // which means max is now included:
+  // [min, max) --> [min, max + 1) AND
+  // Math.floor(Math.random() * (max - min) + min) BECOMES
+  // Math.floor(Math.random() * (max - min + 1) + min).
+  return Math.floor(Math.random() * (maxBound - minBound + 1) + minBound);
```
Comment (kmausolf) [inline pokemon-app/src/utils/randomGenerator.js:39]: The algorithm makes sense, thank you for explaining this, as I would have had a harder time inferring that from the return here.

File: pokemon-app/src/utils/updateHPStats.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * Updates the new HP from the array of the pokemon's stats.
+ * @param {Array} pokeStatsArray
+ * @param {Number} newHP
+ * @param {Boolean} isHPReset Default is false.
+ * @returns {Array} An array with an updated HP.
+ */
+const updateHPStats = (pokeStatsArray, newHP, isHPReset = false) =>
+  pokeStatsArray.map((statsObject) => {
+    // Cloning this to preserve the original.
+    const clonedObject = { ...statsObject };
```
Comment (kmausolf) [inline pokemon-app/src/utils/updateHPStats.js:11]: This works.  You can also use `Object.assign({}, statsObject)` to clone it.

File: pokemon-app/src/components/ProgressBar.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      width: `${progress}%`,
+      background: hpColorStatus,
+    };
+
+    setStyle(newStyle);
+  }, [progress]);
+
+  return (
+    <div className="progress-bar">
+      <span>{Math.round(progress)}%</span>
```
Comment (krferrell) [inline pokemon-app/src/components/ProgressBar.js:26]: Clean progress bar

File: pokemon-app/src/components/SinglePokemon.js [hunk 6]
```diff
@@ -23,29 +26,60 @@ const SinglePokemon = ({
...
+      friendlyPokeMoveTypes[moveIndex]?.generation?.name,
+      friendlyPokeMoveTypes[moveIndex]?.type?.name,
+      friendlyPokeType,
+      enemyPokeStats
+    );
+
+    setEnemyPokeStats(updateHPStats(enemyPokeStats, damageOutcome));
+    setEnemyStatsOnTop(true);
+  };
+
```
Comment (krferrell) [inline pokemon-app/src/components/SinglePokemon.js:63]: This looks legit.

File: pokemon-app/src/components/toad.js [hunk 7]
```diff
@@ -5,13 +5,69 @@ const Toad = {
...
+        damage_class: { name: "special" },
+      },
+    },
+    {
+      move: {
+        name: "Hooks Shot",
+        type: { name: "toad" },
+        power: 9000,
+        generation: { name: "generation-viii" },
+        damage_class: { name: "special" },
```
Comment (krferrell) [inline pokemon-app/src/components/toad.js:23]: This is amazing

**Review by krferrell — APPROVED** [review]
Great work! Cannot wait to see this

File: pokemon-app/src/utils/effectiveness.js [hunk 8]
```diff
@@ -133,14 +140,149 @@ const moveEffectiveness = (defenderType, attackerMoveTypes) => {
...
+      ? (217 * damage) / 255
+      : 0.85 * damage;
+
+  const damageUpperLimit =
+    attackerMoveGeneration === "i" || attackerMoveGeneration === "ii"
+      ? (255 * damage) / 255
+      : 1 * damage;
+
+  // Return the damage, based on luck.
+  return randomGeneratorInclusive(damageLowerLimit, damageUpperLimit);
```
Comment (Lap343) [inline pokemon-app/src/utils/effectiveness.js:285]: this is just insane

Comment (Wolftownn) [reply pokemon-app/src/utils/effectiveness.js:285]: Can't get over how well thought out this is!

**Review by Lap343 — APPROVED** [review]
Very good job implementing the huge damage equation and then having the searched pokemon to have health bars that go down. Again a really good job Elias

**Review by Wolftownn — APPROVED** [review]
I tip my hat to you sir. Incredible job!

**Pull request merged**